### ISSUE TITLE
PHP - pass array(key-value pair) as an argument instead of comma separated individual arguments 

### ISSUE THREAD

**Issue by rajanmodi** [OpenAPITools/openapi-generator#641]
##### Description

Functions generated automatically in SDK should support array(key-value pair) as an argument.

Suppose there is search function with 66 arguments. 
search($this->arg1 = null, $this->arg2 = null, ...., $this->arg66 = null)

When we need to make a call to this search function ,we need to provide all the arguments
  $apiInstance->search($this->arg1 = 5, $this->arg2 = 10, $this->arg3 = 20,.....,$this->arg66 = 1)

What if I only want to pass only two arguments 
 $apiInstance->search($this->arg1 = 5,$this->arg31 = 95).
now this will assign 5 to arg1 and 95 to arg2. I need to assign value of 95 to arg31

I think solution will be generating function argument as an array with key-value pair for such cases . 

Instead of keeping comma separated arguments, array should be there..

**Comment by wing328** [comment]
@rajanmodi thanks for the suggestion.

cc @jebentier @dkarlovi @mandrean @jfastnacht @ackintosh @ybelenko

**Comment by ybelenko** [comment]
@rajanmodi Hmm... Give us an example? I can't find any function with more than one argument in current PHPClient samples. We're talking about models, right?
```php
/**
 * Constructor
 *
 * @param mixed[] $data Associated array of property values
 *                      initializing the model
 */
public function __construct(array $data = null)
{
    $this->container['bill_id'] = isset($data['bill_id']) ? $data['bill_id'] : null;
    $this->container['bill_date'] = isset($data['bill_date']) ? $data['bill_date'] : null;
    $this->container['charges'] = isset($data['charges']) ? $data['charges'] : null;
    $this->container['charge_desc'] = isset($data['charge_desc']) ? $data['charge_desc'] : null;
}
```

**Comment by wing328** [comment]
@ybelenko I think @rajanmodi is referring to method arguments, e.g. [updatePetWithForm](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/2_0/petstore.yaml#L176-L210) with 3 parameters only and let's imagine it has 20 additional optional parameters.

**Comment by ybelenko** [comment]
@wing328 If so, I think that more common practice is to create endpoints with just few parameters, or with one complex JSON object parameter.
I like current "full list of parameters" appouch more and user always can rewrite mustache template a bit.

But key-values pairs very handy when you add both required and optional parameters and rebuild client SDK. You can use fresh SDK right away without need to rearrange arguments in functional call.

In total, 1 pro and 1 con from me :smile:

**Comment by rajanmodi** [comment]
@wing328 <https://github.com/wing328> @ybelenko
<https://github.com/ybelenko> - Yes I am referring to method arguments.
There are 66 arguments in my search method. According to my requirement, to
access search method, I need not pass all the arguments.
I will pass only desired few arguments to search methods and accordingly my
api returns JSON response.

But current SDK code is forcing me to pass all the arguments to make a
search call.

So my suggestion is that instead of keeping comma separated arguments in
search method, it should receive one array(key ,value pair).

So in this case if I need to make a call to search method, I will construct
an array consisting of only needed arguments at my end and pass it to
search method.

On receiving that array, search method will postmortem this array and
accordingly construct those 66 parameters keeping unnecessary argument as
null.

I hope I am able to represent the problem in a concise way. Still In case
of of missing info please do let me know.

Thank you,
Rajan

**Comment by ybelenko** [comment]
@rajanmodi 
66 arguments in search operation, huh?
Even Facebook and Twitter doesn't have API methods with huge arguments list like that.
Anyway, I understand why you don't wanna use object as input parameter, obviosly your HTTP method is `GET` for a search. Maybe you can use enum filters to make arguments list shorter, in way:
```
/jobs/?status_filter=closed|canceled|abandoned
```
@rajanmodi Can you give us endpoint example, when you need huge stack of input parameters?
Another suggestion to solve this problem, we can extend PHPClient config with `useArrayAsArgument` which will be `false` by default.

**Comment by rajanmodi** [comment]
@ybelenko <https://github.com/ybelenko> Basically it is CarSearch Api. You
can refer this documentation for more detail -
https://developer.marketcheck.com/#/Listings/search
https://apidocs.marketcheck.com/

One endpoint is search in which we can pass possible combinations of search
parameters.
for e.g
http://api.marketcheck.com/v1/search?api_key={your_api_key}&latitude=39.73&longitude=-104.99&radius=200&car_type=used&start=0&rows=10&sort_by=dist&sort_order=desc
http://api.marketcheck.com/v1/search?api_key={your_api_key}&latitude=39.73&longitude=-104.99&radius=200&car_type=used&start=0&rows=10&sort_by=dist&sort_order=desc&plot=true
http://api.marketcheck.com/v1/search?api_key={your_api_key}&lease_term=36&lease_down_payment=2000-5000&lease_emp=200-400&sort_by=lease_emp&sort_order=asc

So here I am passing only few parameters and accordingly I will get
response in JSON.

While making a call to search method, I have to pass all 66 parameters to
search, even though I am interested in passing this parameters only -
search($
latitude=39.73,$longitude=-104.99,$radius=200,$car_type=used,$start=0,$rows=10,$sort_by=dist,$sort_order=desc)
rest all arguments I have to pass as null.

**Comment by rajanmodi** [comment]
This solution will work, but I am wandering if there is anything like java adapter classes in php. 
Automatic sdk generation, if supports this java adapter class concept, then this will be great help and  this problem will be solved and great rescue for devs.

**Comment by ackintosh** [comment]
(This is just my personal opinion 😉)

I think that adopting [the adapter class workaround](https://github.com/OpenAPITools/openapi-generator/issues/641#issuecomment-408292155) to SDK generation should avoid as much as possible. 

because...

- [the adapter class](https://github.com/OpenAPITools/openapi-generator/issues/641#issuecomment-408292155) is workaround
- ideally the issue should be resolved with named parameters
- adopting the workaround is easy but removing it is hard as breaking change takes pains to SDK's users.

So I suggest SDK users who has the issue like you reported implement a adapter class like https://github.com/OpenAPITools/openapi-generator/issues/641#issuecomment-408292155 .

**Comment by ackintosh** [comment]
@rajanmodi The group parameter support in PHP client has been implemented by @wing32 and released as v3.3.2. Please give it a try. 😉 

---

#### Samples

- Spec

https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml#L774

- Generated code

https://github.com/OpenAPITools/openapi-generator/blob/master/samples/openapi3/client/petstore/php/OpenAPIClient-php/lib/Api/FakeApi.php#L2538-L2541

**Comment by jeandonaldroselin** [comment]
> @rajanmodi The group parameter support in PHP client has been implemented by @wing328 and released as v3.3.2. Please give it a try. 😉
> 
> #### Samples
> * Spec
> 
> https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/test/resources/3_0/petstore-with-fake-endpoints-models-for-testing.yaml#L774
> 
> * Generated code
> 
> https://github.com/OpenAPITools/openapi-generator/blob/master/samples/openapi3/client/petstore/php/OpenAPIClient-php/lib/Api/FakeApi.php#L2538-L2541

Hello @ackintosh,

First thank you for your great contributions.

I have adopted your tools for a while now but I face the same issue as @rajanmodi.

The consequences of that are multiple : 

* I waste time when writing the methods calling the PHP client methods.

* I do a lot of mistake after updating my PHP client (if I have a parameter which have been added, some method won't work anymore if I forgot to update them)

I don't find any documentation on "how to generate methods supporting array instead of separate arguments as parameters".

Please can you help me ?
